src/solve/fulfill.rs

```diff
@@ -346,7 +346,12 @@ impl<'s> Fulfill<'s> {
     /// the outcome of type inference by updating the replacements it provides.
     pub fn solve(mut self, subst: Substitution) -> Result<Solution> {
         let outcome = self.fulfill()?;
-        if outcome.is_complete() && !self.ambiguous {
+
+        if self.cannot_prove {
```

Comment:
Ah, I see that in this case `unify` would have to return `Outcome` (or `UnifyOutcome`). This way is ok too. It feels a bit less direct to me somehow, but maybe some comments would suffice to clear it up.